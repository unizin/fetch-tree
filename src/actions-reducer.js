export const ACTION_DONE = '@@fetchTree/action_done'
export const STORE_ENTITY = '@@fetchTree/store_entity'

export const setActionStatusDone = (id) => ({
    type: ACTION_DONE,
    payload: id,
})

export const storeEntity = (id, value) => ({
    type: STORE_ENTITY,
    payload: { id, value },
})

export const selectEntity = (state, id) => (
    state.fetchTree != null
    && state.fetchTree.entities != null
    && state.fetchTree.entities[id]
)

export const selectIsReady = (state, id) => Boolean(
    state.fetchTree != null
    && state.fetchTree.ready != null
    && state.fetchTree.ready[id] != null
)


function doneReducer(state = {}, action) {
    if (action.type === 'ACTION_DONE') {
        return {
            ...state,
            [action.payload]: true,
        }
    }
    return state
}

function entityReducer(state = {}, action) {
    if (action.type === STORE_ENTITY) {
        const { id, value } = action.payload
        return {
            ...state,
            [id]: value,
        }
    }
    return state
}


const defaultState = {
    ready: doneReducer(undefined, { type: '__init__' }),
    entities: entityReducer(undefined, { type: '__init__' }),
}

export default function reducer(state = defaultState, action) {
    switch (action.type) {
        case ACTION_DONE:
            return {
                ...state,
                done: doneReducer(state.done, action),
            }
        case STORE_ENTITY:
            return {
                ...state,
                entities: entityReducer(state.done, action),
            }

        default:
            return state
    }
}
