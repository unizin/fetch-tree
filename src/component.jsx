import React from 'react'
import { connect } from 'react-redux'
import loaderContext from './loader_context'
import processor from './processor'
import withProps from './node_types/with-props'
import withDispatch from './node_types/with-dispatch'

const getDisplayName = Component => (
  Component.displayName ||
  Component.name ||
  (typeof Component === 'string' ? Component : 'Component')
)


const contextShape = React.PropTypes.shape({
    execute: React.PropTypes.func.isRequired,
    hasExecuted: React.PropTypes.func.isRequired,
})

// These need to get passed as props and we need to avoid name conflicts with
// any props the user is passing through to Component. Using the dashes and
// colons ensures that we can't accidentally conflict with JSX props.
const IS_READY = '--loader:is-ready'
const ACTION_QUEUE = '--loader:action-queue'
const DISPATCH = '--loader:dispatch'
const DISPATCH_PROXY = '--loader:dispatchProxy'

export default function ({ component: Component, busy: Busy, resources, mapDispatchToProps }) {
    if (mapDispatchToProps) {
        throw new Error('mapDispatchToProps was removed. Use the dispatch node type instead')
    }
    const displayName = `Loader(${getDisplayName(Component)})`

    if (resources.TYPE !== 'group') {
        if (resources.TYPE === 'debug' && resources.child.TYPE === 'group') {
            // It's ok to wrap the group in a debug node
        } else {
            throw new Error('resources must be a `group()`')
        }
    }

    class LoaderComponent extends React.Component {
        static displayName = displayName

        static childContextTypes = {
            loaderContext: contextShape,
        }

        static contextTypes = {
            loaderContext: contextShape,
        }

        constructor(props, context) {
            super(props, context)
            this.loaderContext = loaderContext(context.loaderContext)
            const {
                [DISPATCH]: dispatch,
                [DISPATCH_PROXY]: dispatchProxy,
            } = props
            dispatchProxy.dispatch = dispatch
        }

        getChildContext() {
            return {
                loaderContext: this.loaderContext,
            }
        }

        componentDidMount() {
            const {
                [ACTION_QUEUE]: actionQueue,
                [DISPATCH]: dispatch,
            } = this.props
            this.loaderContext.execute(dispatch, actionQueue)
        }

        componentWillReceiveProps(nextProps) {
            const {
                [DISPATCH]: dispatch,
                [DISPATCH_PROXY]: dispatchProxy,
            } = nextProps
            dispatchProxy.dispatch = dispatch
        }

        componentDidUpdate() {
            const {
                [ACTION_QUEUE]: actionQueue,
                [DISPATCH]: dispatch,
            } = this.props
            this.loaderContext.execute(dispatch, actionQueue)
        }

        render() {
            const {
                [IS_READY]: isReady,
                [ACTION_QUEUE]: ignoredActionQueue,
                [DISPATCH_PROXY]: ignoredDispatchProxy,
                ...props,
            } = this.props

            if (!isReady) {
                if (Busy) {
                    return (<Busy />)
                }

                return (<noscript></noscript>)
            }

            return (
                <Component {...props} />
            )
        }
    }

    function mapStateToProps() {
        // Each instance gets its own `dispatchProxy` function that never changes
        const dispatchProxy = (action) => {
            // dispatchProxy.dispatch will get attached in the component's constructor
            // and updated in componentWillReceiveProps
            if (!dispatchProxy.dispatch) {
                throw new Error(`Dispatch can't be called yet`)
            }
            return dispatchProxy.dispatch(action)
        }

        return (state, props) => {
            const localResources = withProps(props,
                withDispatch(dispatchProxy, resources)
            )
            const { isReady, actionQueue, value } = processor(localResources, state)

            return {
                ...value,
                [IS_READY]: isReady,
                [ACTION_QUEUE]: actionQueue,
                [DISPATCH_PROXY]: dispatchProxy,
            }
        }
    }

    const mapDispatch = {
        [DISPATCH]: (action) => action,
    }

    return connect(mapStateToProps, mapDispatch)(LoaderComponent)
}
