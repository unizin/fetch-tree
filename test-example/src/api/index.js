import * as api from './api'

let counts = {}

Object.keys(api).reduce((api, k) => {
    const implementation = api[k]

    api[k] = function apiCounter(...args) {
        counts[k] = 1 + (counts[k] || 0)
        return implementation.apply(this, args)
    }

    return api
}, api)


export const resetCounts = () => { counts = {} }
export const getCounts = () => counts

export default api
