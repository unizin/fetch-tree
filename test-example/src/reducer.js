import { combineReducers } from 'redux'
import { reducer as fetchTree } from 'fetch-tree'

function todos(state = {}, action) {
    switch (action.type) {
        case 'SET_TODO': {
            return {
                ...state,
                [action.meta.id]: action.payload,
            }
        }
        default:
            return state
    }
}

function count(state = 0, action) {
    switch (action.type) {
        case 'INCREMENT_COUNT':
            return state + 1
        default:
            return state
    }
}

export default combineReducers({
    currentUsername: 'bob',
    todos,
    count,
    fetchTree,
})


export const selectTodo = (state, id) => state.todos[id]

export const selectCount = (state) => state.count

export const selectCurrentUsername = (state) => state.currentUsername
