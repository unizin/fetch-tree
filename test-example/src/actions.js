import api from './api'

export const fetchTodo = (id) => (dispatch) => {
    return api.fetchTodo(id).then(todo => {
        dispatch({
            type: 'SET_TODO',
            payload: todo,
            meta: { id },
        })
    })
}

export const incrementCount = () => ({
    type: 'INCREMENT_COUNT',
})
