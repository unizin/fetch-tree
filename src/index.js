import group from './node_types/group'
import loader from './node_types/loader'
import selector from './node_types/selector'
import depends from './node_types/depends'
import debug from './node_types/debug'
import virtual from './node_types/virtual'
import entity from './node_types/entity'
import withContext from './node_types/with-context'
import child from './node_types/child'
import withProps from './node_types/with-props'
import withDispatch from './node_types/with-dispatch'
import dispatch from './node_types/dispatch'
import fromProps from './node_types/from-props'

import component from './component'
import reducer, { refreshCache } from './actions-reducer'

import { register as registerNodeType } from './processor.js'

export { group, loader, selector, depends, reducer, debug, virtual, entity, withContext }
export { withDispatch, dispatch }
export { child, withProps, fromProps }
export { refreshCache }

export { registerNodeType }
export default component
