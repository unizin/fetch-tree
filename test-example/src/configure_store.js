import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import reducer from './reducer'

export default function configureStore() {
    const store = createStore(
        reducer,
        undefined,
        applyMiddleware(
            thunk
        )
    )

    return store
}
