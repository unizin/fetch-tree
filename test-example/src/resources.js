import { loader, entity } from 'fetch-tree'
import { fetchTodo } from './actions'
import { selectTodo } from './reducer'
import api from './api'

export const todoResource = loader({
    id: (id) => `todo-${id}`,
    action: fetchTodo,
    selector: selectTodo,
})

export const branchResource = entity({
    id: (id) => `branch-${id}`,
    apiFunction: api.fetchBranch,
})

export const userResource = entity({
    id: (id) => `user-${id}`,
    apiFunction: api.fetchUser,
})
