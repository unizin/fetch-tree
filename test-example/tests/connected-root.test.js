import React from 'react'
import renderer from 'react-test-renderer'

import { renderSnapshot } from './test-utils.js'
import fetchTree, { group, fromProps, child, depends } from 'fetch-tree'

import Loading from '../src/components/loading'

import { branchResource } from '../src/resources'
// These are FetchTree components, but are not connected
import LeafA from '../src/components/leaf-a'
import LeafB from '../src/components/leaf-b'
import Todo from '../src/components/todo'

/*
Read the connected-leaves tests first
*/

function Branch(props) {
    return (
        <div>
            {`Branch props: ${JSON.stringify(props, undefined, 2)}`}
            {/*

                You can't just pass id here because these leaves aren't
                connected and can't pull its data from the store.

            */}
            <h1>{props.branch.name}</h1>
            <LeafA {...props.leafAProps} />
            <LeafB {...props.leafBProps} />
        </div>
    )
}
Branch.propTypes = {
    branch: React.PropTypes.shape({
        // This has other props, but this component doesn't care about them
        name: React.PropTypes.string.isRequired,
    }).isRequired,
    // Branch doesn't know or care what the shape is
    leafAProps: React.PropTypes.shape(
        LeafA.propTypes
    ).isRequired,
    leafBProps: React.PropTypes.shape(
        LeafB.propTypes
    ).isRequired,
}
const ConnectedBranch = fetchTree({
    component: Branch,
    busy: Loading,
    resourceGroup: group({
        id: fromProps('id'),
        branch: depends(
            ['id'],
            branchResource
        ),
        // this child() is not wrapped in virtual. it's going to gether the
        // props that need to be passed down and store them in `leafAProps`
        leafAProps: child(
            // When using child() or depends() you may depend on any named
            // resource defined from above
            ['branch.leafAId', 'id'],
            (id, branchId) => <LeafA id={id} branchId={branchId} />
        ),
        leafBProps: child(
            ['branch.leafBId'],
            (id) => <LeafB id={id} />
        ),
    }),
})

test(`Connected root composed of unconnected branches and leaves`, async () => {
    const report = await renderSnapshot(
        <ConnectedBranch id={1} />
    )
    expect(report.apiRequests).toEqual({
        fetchBranch: 1,
        fetchTodo: 2,
    })
    expect(report.loadingScreens).toBe(1)
})

test(`Connected root with an array of unconnected children`, async () => {
    let actualProps
    function Root(props) {
        actualProps = props
        return (
            <div>
                {props.todoProps.map(
                    (props, index) => <Todo key={index} {...props} />
                )}
            </div>
        )
    }
    const ConnectedRoot = fetchTree({
        component: Root,
        busy: Loading,
        resourceGroup: group({
            ids: fromProps('ids'),
            todoProps: child(
                ['ids'],
                ids => ids.map(
                    id => <Todo id={id} />
                )
            ),
        }),
    })
    const report = await renderSnapshot(
        <ConnectedRoot ids={[1, 2, 3]} />
    )
    expect(actualProps.ids.length).toBe(actualProps.todoProps.length)
    expect(report.apiRequests).toEqual({
        fetchTodo: 3,
    })
    expect(report.loadingScreens).toBe(1)
})

test(`Disconnected root with mock/fixture props`, async () => {
    const branch = { name: 'Branch Fixture' }
    const leafA = { id: 11, note: 'LeafA fixture' }
    const leafB = { id: 42, note: 'LeafA fixture' }

    const leafAProps = { data: leafA }
    const leafBProps = { data: leafB }

    const component = renderer.create(
        <Branch branch={branch} leafAProps={leafAProps} leafBProps={leafBProps} />
    )
    // see __snapshots__ for the the HTML the component produced.
    expect(component.toJSON()).toMatchSnapshot()
})
