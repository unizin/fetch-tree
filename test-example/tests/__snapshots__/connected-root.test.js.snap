exports[`test Connected root composed of unconnected branches and leaves 1`] = `
<div>
  Branch props: {
    "id": 1,
    "branch": {
      "id": 1,
      "name": "nepik",
      "leafAId": 83,
      "leafBId": 31
    },
    "leafAProps": {
      "id": 83,
      "branchId": 1,
      "data": {
        "id": 83,
        "done": true,
        "note": "Akitunca materu ihu gi sew mile kacadifes vo pi nemukla vedpuunu bo huicif ubo cerafeip woswe bi upe."
      }
    },
    "leafBProps": {
      "id": 31,
      "data": {
        "id": 31,
        "done": false,
        "note": "Cupufsah foma haki nobdiv ze temaf la vapatukop gozgisi ham uzejagiso gotsejloj wepup zoselje."
      }
    }
  }
  <h1>
    nepik
  </h1>
  <div>
    LeafA props: {
      "id": 83,
      "branchId": 1,
      "data": {
        "id": 83,
        "done": true,
        "note": "Akitunca materu ihu gi sew mile kacadifes vo pi nemukla vedpuunu bo huicif ubo cerafeip woswe bi upe."
      }
    }
  </div>
  <div>
    LeafB props: {
      "id": 31,
      "data": {
        "id": 31,
        "done": false,
        "note": "Cupufsah foma haki nobdiv ze temaf la vapatukop gozgisi ham uzejagiso gotsejloj wepup zoselje."
      }
    }
  </div>
</div>
`;

exports[`test Connected root with an array of unconnected children 1`] = `
<div>
  <div>
    id: 1
    <input
      checked={false}
      type="checkbox" />
    <p>
      Fouti hi rovumjak wuwab ojki utemun ti voblom mefup ivkiz nadahuw ci.
    </p>
  </div>
  <div>
    id: 2
    <input
      checked={true}
      type="checkbox" />
    <p>
      Bev rip asu rosob lo su za wirun kubvaw nadridzez cabfed temekuinu co.
    </p>
  </div>
  <div>
    id: 3
    <input
      checked={true}
      type="checkbox" />
    <p>
      Liuna igana pucodmek emik hapo tuczuc vokme podikru vokuse horu mum tion catke pohabe wipewen zavse osawa.
    </p>
  </div>
</div>
`;

exports[`test Disconnected root with mock/fixture props 1`] = `
<div>
  Branch props: {
    \"branch\": {
      \"name\": \"Branch Fixture\"
    },
    \"leafAProps\": {
      \"data\": {
        \"id\": 11,
        \"note\": \"LeafA fixture\"
      }
    },
    \"leafBProps\": {
      \"data\": {
        \"id\": 42,
        \"note\": \"LeafA fixture\"
      }
    }
  }
  <h1>
    Branch Fixture
  </h1>
  <div>
    LeafA props: {
      \"data\": {
        \"id\": 11,
        \"note\": \"LeafA fixture\"
      }
    }
  </div>
  <div>
    LeafB props: {
      \"data\": {
        \"id\": 42,
        \"note\": \"LeafA fixture\"
      }
    }
  </div>
</div>
`;
